### PR TITLE
Marking move from one component type to another one as impossible for schema evolution

### PR DESCRIPTION
**Pull request by hegner**
BEGINRELEASENOTES
- Detect more impossible schema evolutions
- Improving printout of schema evolution checker
ENDRELEASENOTES

This addresses #563

### CODE REVIEW

File: python/podio_schema_evolution.py [hunk 1]
```diff
@@ -105,8 +105,8 @@ def __init__(self, name, member_name, old_member, new_member):
         self.new_member = new_member
         self.klassname = name
         super().__init__(
-            f"'{self.name}.{self.member_name}' changed type from '+\
-                    '{self.old_member.full_type} to {self.new_member.full_type}"
+            f"'{self.name}.{self.member_name}' changed type from "+\
+                    f"{self.old_member.full_type} to {self.new_member.full_type}"
```
Comment (tmadlener) [inline python/podio_schema_evolution.py:109]:
```suggestion
            f"'{self.name}.{self.member_name}' changed type from "
            f"{self.old_member.full_type} to {self.new_member.full_type}"
```

File: python/podio_schema_evolution.py [hunk 2]
```diff
@@ -469,4 +475,4 @@ def read_evolution_file(self) -> None:
     comparator.read()
     comparator.compare()
     comparator.print_comparison()
-    print(comparator.get_changed_schemata(schema_filter=root_filter))
+    #print(comparator.get_changed_schemata(schema_filter=root_filter))
```
Comment (tmadlener) [inline python/podio_schema_evolution.py:478]: I suppose this was some debug output at some point, which is why it is removed?

**Review by tmadlener — COMMENTED** [review]
This turned out to be a slightly simpler change than what I would have imagined :).

File: python/podio_schema_evolution.py (hunk 2 above)
Comment (hegner) [reply python/podio_schema_evolution.py]: sort of. still useful for the future to include again

**Comment by hegner** [comment]
> This turned out to be a slightly simpler change than what I would have imagined :).

Probably speaks for the design for once :-D

**Review by tmadlener — COMMENTED** [review]
There seems to be an unrelated error that currently makes the script fail when trying to run it on the edm4hep.yaml file that triggered the original issue:

```console
Traceback (most recent call last):
  File "/home/tmadlener/work/AIDASoft/podio/python/podio_schema_evolution.py", line 476, in <module>
    comparator.compare()
  File "/home/tmadlener/work/AIDASoft/podio/python/podio_schema_evolution.py", line 184, in compare
    self._compare_components()
  File "/home/tmadlener/work/AIDASoft/podio/python/podio_schema_evolution.py", line 196, in _compare_components
    [
  File "/home/tmadlener/work/AIDASoft/podio/python/podio_schema_evolution.py", line 197, in <listcomp>
    AddedComponent(self.datamodel_new.components[name], name)
  File "/home/tmadlener/work/AIDASoft/podio/python/podio_schema_evolution.py", line 34, in __init__
    super().__init__(f"'{self.component.name}' has been added")
AttributeError: 'dict' object has no attribute 'name'
```

The issue is here:
https://github.com/AIDASoft/podio/blob/ab6b2a4374709d98407102e841edb9690b04d039/python/podio_schema_evolution.py#L34

it should be `self.name` instead of `self.component.name` to make things work again. With that fix the script detects invalid schema changes in the EDM4hep yaml file.

Can we add a simple test that at least runs the script on the example datamodel(s) that we have?

**Comment by hegner** [comment]
> Can we add a simple test that at least runs the script on the example datamodel(s) that we have?
Surprised we didn't have. This PR or a new one?

**Comment by tmadlener** [comment]
Yeah, I also thought we had one, but at least I couldn't find one. Or if we have one at least it didn't trigger this issue. I would prefer a new PR.

**Comment by hegner** [comment]
We only had for the C++ part it seems. OK. new ticket, new PR

**Comment by tmadlener** [comment]
PR to fix the bug in the schema evolution generation: #627

**Comment by hegner** [comment]
other PR merged now

**Comment by tmadlener** [comment]
Can you rebase this quickly to pick up those changes?

**Comment by hegner** [comment]
obviously that fails now ;-) need to change the test definition

**Review by tmadlener — COMMENTED** [review]
Works for me. Ideally we could condense down the test for a failing schema evolution a bit rather than having the third copy of the datamodel effectively, but we can also postpone that to after v1.0

**Comment by hegner** [comment]
for post v1 we could do that a bit separated without a file, but in pure python tests. I will merge now.

**Pull request merged**